packages/apollo-server-hapi/src/ApolloServer.ts

```diff
@@ -90,66 +93,97 @@ export class ApolloServer extends ApolloServerBase {
+            query:
+              request.method === 'post'
+                ? // TODO type payload as string or Record
+                  (request.payload as any)
+                : request.query,
+            request: convertNodeHttpToRequest(request.raw.req),
+          },
+        );
+
+        const response = h.response(graphqlResponse);
+        if (responseInit.headers) {
+          Object.entries(responseInit.headers).forEach(([headerName, value]) =>
+            response.header(headerName, value),
           );
+        }
+        response.code(responseInit.status || 200);
+        return response;
+      } catch (error) {
+        if (!isHttpQueryError(error)) {
+          throw Boom.boomify(error as Error);
+        }
 
-          const response = h.response(graphqlResponse);
-          if (responseInit.headers) {
-            Object.entries(responseInit.headers).forEach(
-              ([headerName, value]) => response.header(headerName, value),
-            );
-          }
-          response.code(responseInit.status || 200);
-          return response;
-        } catch (error) {
-          if (!isHttpQueryError(error)) {
-            throw Boom.boomify(error as Error);
-          }
-
-          if (true === error.isGraphQLError) {
-            const response = h.response(error.message);
-            if (error.headers) {
-              Object.entries(error.headers).forEach(([headerName, value]) => {
-                response.header(headerName, value);
-              });
-            }
-            response.code(error.statusCode);
-            response.type('application/json');
-            return response;
-          }
-
-          const err = new Boom.Boom(error.message, {
-            statusCode: error.statusCode,
-          });
+        if (true === error.isGraphQLError) {
+          const response = h.response(error.message);
           if (error.headers) {
             Object.entries(error.headers).forEach(([headerName, value]) => {
-              err.output.headers[headerName] = value;
+              response.header(headerName, value);
             });
           }
-          // Boom hides the error when status code is 500
-          err.output.payload.message = error.message;
-          throw err;
+          response.code(error.statusCode);
+          response.type('application/json');
+          return response;
         }
-      },
+
+        const err = new Boom.Boom(error.message, {
+          statusCode: error.statusCode,
+        });
+        if (error.headers) {
+          Object.entries(error.headers).forEach(([headerName, value]) => {
+            err.output.headers[headerName] = value;
+          });
+        }
+        // Boom hides the error when status code is 500
+        err.output.payload.message = error.message;
+        throw err;
+      }
+    };
+
+    // POST ROUTE
+    const postOptions = { ...route, ...routePost };
+
+    // if we have post route options which contain no cors options, but have
```

Comment:
This doesn't look like a backwards compatible change. It looks like previously, if you pass both `route` and `cors`, the `cors` argument would be ignored. That doesn't seem like a great choice, but we should avoid changing existing behavior unless there's a really good reason.

Reply:
Yeah, I see what you are saying. It's unfortunate that you would have to be duplicative in CORS settings if you provide any route options at all. Technically, there may be some cases where someone relied on the route information wiping out the CORS settings, but I would think this is likely to be exceedingly rare.

I'd say that being able to expect the passed in CORS settings to work is a pretty good and reasonable breaking change in this case. What's the alternative to simply documenting this as a potentially breaking change? Provide a feature flag that can be passed in enable the "new" behavior that behaves more as one would expect?

Reply:
I'd like to avoid breaking changes in what is likely to be our last ever change to apollo-server-hapi as a core-maintained package versioned identically with the rest of the package (https://github.com/apollographql/apollo-server/issues/6028). Once apollo-server-hapi is a separate package which can bump major versions independently of the new `@apollo/server`, such changes would be more reasonable.

Reply:
Fair enough. I'll unwind this.

Reply:
Sorry it's been so long getting back to this. I'm checking in a change, which sets default cors options as per before only when route is not specified. However, with the new routeGet and routePost, the cors options are merged in (as these are new and would not affect backward compatibility)

Reply:
The change should now not be breaking. The CORs merging only happens with the new get / post route options parameters are used and the same behavior should exist for the standard route option (including defaulting cors when not present)